```
FAILED tests/selenium/test_quizzes_page_ui.py::test_completed_quiz_modal_flow - assert 'active' in 'quiz-overlay'
FAILED tests/selenium/test_quizzes_page_ui.py::test_sections_update_after_submission - selenium.common.exceptions.TimeoutException: Message:
```

--- a/tests/selenium/test_quizzes_page_ui.py
+++ b/tests/selenium/test_quizzes_page_ui.py
@@ -209,3 +209,50 @@ def test_todo_quiz_full_flow(driver):
     )
     assert driver.find_element(By.CSS_SELECTOR, ".modal-results").is_displayed()
     assert len(driver.find_elements(By.CSS_SELECTOR, ".keyword-chip")) == 2
+
+#Finally we check that the sections update correctly after a quiz submission
+def test_sections_update_after_submission(driver):
+    login_user(driver)
+    go_to_quizzes(driver)
+
+    WebDriverWait(driver, 10).until(
+        EC.presence_of_element_located((By.CSS_SELECTOR, "#todo-list .quiz-card"))
+    )
+
+    #We record the state of the progress bar and completed quiz card numbers to have as a comparison point
+    initial_progress = driver.find_element(By.ID, "overall-progress-percent").text
+    initial_completed_count = len(driver.find_elements(By.CSS_SELECTOR, "#completed-list .quiz-card"))
+
+    #We find and click the time availability quiz card
+    todo_cards = driver.find_elements(By.CSS_SELECTOR, "#todo-list .quiz-card")
+    time_availability_card = next(
+        card for card in todo_cards
+        if "Time Availability" in card.find_element(By.CSS_SELECTOR, ".quiz-card-name").text
+    )
+    time_availability_card.click()
+    time.sleep(1)
+
+    #We complete the quiz by calling the helper function
+    driver.find_element(By.ID, "begin-btn").click()
+    time.sleep(1)
+    answer_all_questions(driver)
+
+    WebDriverWait(driver, 10).until(
+        EC.presence_of_element_located((By.CSS_SELECTOR, ".modal-results"))
+    )
+
+    #We close the modal
+    close_modal(driver)
+    time.sleep(2)
+
+    #We check the progress bar percentage has increased
+    assert driver.find_element(By.ID, "overall-progress-percent").text != initial_progress
+
+    #THen we check the completed section has one more card
+    updated_completed_count = len(driver.find_elements(By.CSS_SELECTOR, "#completed-list .quiz-card"))
+    assert updated_completed_count == initial_completed_count + 1
+
+    #We check that there are no duplicate cards anywhere on the page
+    all_cards = driver.find_elements(By.CSS_SELECTOR, ".quiz-card")
+    card_names = [c.find_element(By.CSS_SELECTOR, ".quiz-card-name").text for c in all_cards]
+    assert len(card_names) == len(set(card_names))

--- a/tests/selenium/test_quizzes_page_ui.py
+++ b/tests/selenium/test_quizzes_page_ui.py
@@ -166,3 +166,46 @@ def test_completed_quiz_modal_flow(driver):
     #And check that the modal is truly closed
     assert "active" not in driver.find_element(By.ID, "quiz-overlay").get_attribute("class")
 
+#We then test the full flow for a todo quiz card
+def test_todo_quiz_full_flow(driver):
+    login_user(driver)
+    go_to_quizzes(driver)
+
+    #We wait until a quiz card loads into the to-do section
+    WebDriverWait(driver, 10).until(
+        EC.presence_of_element_located((By.CSS_SELECTOR, "#todo-list .quiz-card"))
+    )
+
+    #We look for the time availability quiz card as charlie hasnt done it yet
+    todo_cards = driver.find_elements(By.CSS_SELECTOR, "#todo-list .quiz-card")
+    time_availability_card = next(
+        card for card in todo_cards
+        if "Time Availability" in card.find_element(By.CSS_SELECTOR, ".quiz-card-name").text
+    )
+
+    #We then open the selected quiz card
+    time_availability_card.click()
+    time.sleep(1)
+
+    #We check that the intro modal slide is opened with the begin button
+    assert driver.find_element(By.CSS_SELECTOR, ".modal-intro").is_displayed()
+    begin_btn = driver.find_element(By.ID, "begin-btn")
+    assert begin_btn.is_displayed()
+
+    #We then click the begin button
+    begin_btn.click()
+    time.sleep(1)
+
+    #And check that the question modal slide appears with the slider
+    assert driver.find_element(By.CSS_SELECTOR, ".modal-question").is_displayed()
+    assert driver.find_element(By.ID, "q-slider").is_displayed()
+
+    #We call the helper function to answer all 10 questions
+    answer_all_questions(driver)
+
+    #We then confirm the results slide appears with exactly 2 keyword chips
+    WebDriverWait(driver, 10).until(
+        EC.presence_of_element_located((By.CSS_SELECTOR, ".modal-results"))
+    )
+    assert driver.find_element(By.CSS_SELECTOR, ".modal-results").is_displayed()
+    assert len(driver.find_elements(By.CSS_SELECTOR, ".keyword-chip")) == 2

--- a/tests/selenium/test_quizzes_page_ui.py
+++ b/tests/selenium/test_quizzes_page_ui.py
@@ -104,3 +104,65 @@ def test_quizzes_page_loads_correctly(driver):
     #We also check that the overall progress bar is visible and shows the progress text
     assert driver.find_element(By.CSS_SELECTOR, ".overall-progress-section").is_displayed()
     assert "quizzes completed" in driver.find_element(By.ID, "overall-progress-text").text
+
+#We then test the modal flow for when we click on a quiz card in the completed section
+def test_completed_quiz_modal_flow(driver):
+    #We use our helper functions to login and navigate to the page
+    login_user(driver)
+    go_to_quizzes(driver)
+
+    #We then wait until the quiz cards appear in the completed list section
+    WebDriverWait(driver, 10).until(
+        EC.presence_of_element_located((By.CSS_SELECTOR, "#completed-list .quiz-card"))
+    )
+
+    #We then click on a completed quiz card
+    driver.find_elements(By.CSS_SELECTOR, "#completed-list .quiz-card")[0].click()
+    time.sleep(1)
+
+    #A modal should then open and we confirm that it does and that 2 buttons are present, retake and view results
+    assert "active" in driver.find_element(By.ID, "quiz-overlay").get_attribute("class")
+    assert driver.find_element(By.CSS_SELECTOR, ".btn-retake").is_displayed()
+    assert driver.find_element(By.CSS_SELECTOR, ".btn-view-results").is_displayed()
+
+    #We first check if the view results button works but clicking on it
+    driver.find_element(By.CSS_SELECTOR, ".btn-view-results").click()
+    time.sleep(1)
+
+    #We then check to see if the key words are displayed properly
+    assert driver.find_element(By.CSS_SELECTOR, ".modal-results").is_displayed()
+    assert len(driver.find_elements(By.CSS_SELECTOR, ".keyword-chip")) > 0
+
+    #Next we close the modal
+    close_modal(driver)
+    time.sleep(1)
+
+    #And click to reopen the exact same quiz card and check the other button option
+    driver.find_elements(By.CSS_SELECTOR, "#completed-list .quiz-card")[0].click()
+    time.sleep(1)
+
+    #We click the other button to retake the quiz
+    driver.find_element(By.CSS_SELECTOR, ".btn-retake").click()
+    time.sleep(1)
+
+    #We then check that a modal opens and that it is the question modal as we move to retake the quiz
+    assert driver.find_element(By.CSS_SELECTOR, ".modal-question").is_displayed()
+    #We also check to see if the question slider appears
+    assert driver.find_element(By.ID, "q-slider").is_displayed()
+
+    #Next we call the helper function to complete all the quiz questions
+    answer_all_questions(driver)
+
+    #After the quiz is finished we then confirm that the results modal slide is showed
+    WebDriverWait(driver, 10).until(
+        EC.presence_of_element_located((By.CSS_SELECTOR, ".modal-results"))
+    )
+    assert driver.find_element(By.CSS_SELECTOR, ".modal-results").is_displayed()
+
+    #We then close the modal
+    close_modal(driver)
+    time.sleep(1)
+
+    #And check that the modal is truly closed
+    assert "active" not in driver.find_element(By.ID, "quiz-overlay").get_attribute("class")
+

